// antigravity-ide/cli/manifest.js
/**
 * Antigravity Resource Manifest
 * Defines the whitelist of agents, rules, and workflows for each operation mode.
 */

const MANIFEST = {
    eco: {
        rules: ['GEMINI.md', 'security.md'],
        agents: [
            'orchestrator.md',
            'project-planner.md',
            'frontend-specialist.md',
            'backend-specialist.md',
            'debugger.md'
        ],
        workflows: [
            'create.md',
            'plan.md',
            'debug.md',
            'orchestrate.md',
            'status.md',
            'log-error.md'
        ],
        skills: [
            'ab-test-setup',
            'address-github-comments',
            'agent-backend-patterns',
            'agent-code-guide',
            'agent-frontend-patterns',
            'agent-orchestration',
            'agent-project-guidelines-example',
            'airflow-dag-patterns',
            'api-documentation-generator',
            'api-documenter',
            'api-patterns',
            'architecture-patterns',
            'async-python-patterns',
            'auth-implementation-patterns',
            'autonomous-agent-patterns',
            'backend-dev-guidelines',
            'bash-defensive-patterns',
            'bash-linux',
            'bash-pro',
            'bats-testing-patterns',
            'binary-analysis-patterns',
            'brand-guidelines-community',
            'c4-context',
            'cicd-automation-workflow-automate',
            'clean-code',
            'code-documentation-code-explain',
            'code-documentation-doc-generate',
            'code-refactoring-context-restore',
            'concise-planning',
            'conductor-setup',
            'context-driven-development',
            'context-management-context-restore',
            'context-management-context-save',
            'context-manager',
            'context-window-management',
            'context7-auto-research',
            'dbt-transformation-patterns',
            'debugger',
            'debugging-strategies',
            'debugging-toolkit-smart-debug',
            'defi-protocol-templates',
            'distributed-debugging-debug-trace',
            'doc-coauthoring',
            'docker-expert',
            'docs-architect',
            'documentation-generation-doc-generate',
            'documentation-templates',
            'docx-official',
            'dotnet-backend-patterns',
            'e2e-testing-patterns',
            'employment-contract-templates',
            'environment-setup-guide',
            'error-debugging-error-analysis',
            'error-debugging-error-trace',
            'error-debugging-multi-agent-review',
            'error-diagnostics-smart-debug',
            'error-handling-patterns',
            'executing-plans',
            'fastapi-templates',
            'frontend-dev-guidelines',
            'frontend-mobile-development-component-scaffold',
            'full-stack-scaffold',
            'git-advanced-workflows',
            'git-collaboration-master',
            'git-pr-workflows-git-workflow',
            'git-pr-workflows-onboard',
            'git-pr-workflows-pr-enhance',
            'git-pushing',
            'github-actions-templates',
            'github-mcp',
            'github-workflow-automation',
            'gitlab-ci-patterns',
            'gitops-workflow',
            'go-concurrency-patterns',
            'godot-gdscript-patterns',
            'helm-chart-scaffolding',
            'incident-runbook-templates',
            'javascript-testing-patterns',
            'javascript-typescript-typescript-scaffold',
            'linkerd-patterns',
            'llm-app-patterns',
            'memory-safety-patterns',
            'microservices-patterns',
            'ml-pipeline-workflow',
            'modern-javascript-patterns',
            'nextjs-app-router-patterns',
            'nodejs-backend-patterns',
            'notion-template-business',
            'nx-workspace-patterns',
            'observability-monitoring-monitor-setup',
            'obsidian-clipper-template-creator',
            'on-call-handoff-patterns',
            'plan-writing',
            'planning-with-files',
            'powershell-windows',
            'projection-patterns',
            'prompt-engineering-patterns',
            'python-development-python-scaffold',
            'python-patterns',
            'python-testing-patterns',
            'react-patterns',
            'react-ui-patterns',
            'rust-async-patterns',
            'seo-content-planner',
            'seo-structure-architect',
            'similarity-search-patterns',
            'sql-optimization-patterns',
            'startup-analyst',
            'startup-business-analyst-business-case',
            'startup-business-analyst-financial-projections',
            'startup-business-analyst-market-opportunity',
            'startup-financial-modeling',
            'startup-metrics-framework',
            'stride-analysis-patterns',
            'systematic-debugging',
            'tailwind-patterns',
            'tdd-master-workflow',
            'tdd-workflow',
            'tdd-workflows-tdd-cycle',
            'tdd-workflows-tdd-green',
            'tdd-workflows-tdd-red',
            'tdd-workflows-tdd-refactor',
            'testing-patterns',
            'unity-ecs-patterns',
            'using-git-worktrees',
            'wcag-audit-patterns',
            'web-design-guidelines',
            'workflow-automation',
            'workflow-orchestration-patterns',
            'workflow-patterns',
            'writing-plans',
            'zapier-make-patterns'
        ],
        shared: [
            'i18n-master',
            'design-philosophy',
            'ai-master'
        ],
        core: '*', // Personality & Archetypes
        scripts: [
            'checklist.py',
            'verify_all.py'
        ]
    },
    pro: {
        rules: [
            'GEMINI.md',
            'security.md',
            'frontend.md',
            'backend.md',
            'debug.md'
        ],
        agents: [
            'orchestrator.md',
            'project-planner.md',
            'frontend-specialist.md',
            'backend-specialist.md',
            'debugger.md',
            'devops-architect.md',
            'documentation-writer.md',
            'test-engineer.md',
            'codebase-expert.md',
            'security-auditor.md'
        ],
        workflows: [
            'create.md',
            'plan.md',
            'debug.md',
            'orchestrate.md',
            'status.md',
            'test.md',
            'deploy.md',
            'monitor.md',
            'audit.md',
            'document.md',
            'log-error.md'
        ],
        skills: [
            'ab-test-setup',
            'accessibility-compliance-accessibility-audit',
            'address-github-comments',
            'agent-backend-patterns',
            'agent-code-guide',
            'agent-frontend-patterns',
            'agent-orchestration',
            'agent-project-guidelines-example',
            'agent-security-review',
            'ai-engineer',
            'airflow-dag-patterns',
            'api-design-principles',
            'api-documentation-generator',
            'api-documenter',
            'api-fuzzing-bug-bounty',
            'api-patterns',
            'api-security-best-practices',
            'api-testing-observability-api-mock',
            'app-store-optimization',
            'application-performance-performance-optimization',
            'architect-review',
            'architecture',
            'architecture-decision-records',
            'architecture-patterns',
            'async-python-patterns',
            'auth-implementation-patterns',
            'autonomous-agent-patterns',
            'aws-penetration-testing',
            'backend-architect',
            'backend-dev-guidelines',
            'backend-development-feature-development',
            'backend-security-coder',
            'backtesting-frameworks',
            'bash-defensive-patterns',
            'bash-linux',
            'bash-pro',
            'bats-testing-patterns',
            'bazel-build-optimization',
            'binary-analysis-patterns',
            'brand-guidelines-community',
            'broken-authentication',
            'bullmq-specialist',
            'burp-suite-testing',
            'c4-architecture-c4-architecture',
            'c4-context',
            'canvas-design',
            'cicd-automation-workflow-automate',
            'clean-code',
            'clerk-auth',
            'cloud-architect',
            'cloud-architect-master',
            'cloud-penetration-testing',
            'code-documentation-code-explain',
            'code-documentation-doc-generate',
            'code-refactoring-context-restore',
            'code-review-ai-ai-review',
            'code-review-checklist',
            'code-review-excellence',
            'code-reviewer',
            'codebase-cleanup-deps-audit',
            'codex-review',
            'comprehensive-review-full-review',
            'comprehensive-review-pr-enhance',
            'concise-planning',
            'conductor-setup',
            'context-driven-development',
            'context-management-context-restore',
            'context-management-context-save',
            'context-manager',
            'context-window-management',
            'context7-auto-research',
            'cost-optimization',
            'data-quality-frameworks',
            'data-scientist',
            'database-admin',
            'database-architect',
            'database-cloud-optimization-cost-optimize',
            'database-design',
            'database-migration',
            'database-migrations-migration-observability',
            'database-migrations-sql-migrations',
            'database-optimizer',
            'dbt-transformation-patterns',
            'debugger',
            'debugging-strategies',
            'debugging-toolkit-smart-debug',
            'defi-protocol-templates',
            'dependency-management-deps-audit',
            'deployment-engineer',
            'deployment-pipeline-design',
            'deployment-procedures',
            'deployment-validation-config-validate',
            'design-orchestration',
            'devops-troubleshooter',
            'distributed-debugging-debug-trace',
            'distributed-tracing',
            'doc-coauthoring',
            'docker-expert',
            'docs-architect',
            'documentation-generation-doc-generate',
            'documentation-templates',
            'docx-official',
            'dotnet-backend-patterns',
            'e2e-testing-patterns',
            'employment-contract-templates',
            'environment-setup-guide',
            'error-debugging-error-analysis',
            'error-debugging-error-trace',
            'error-debugging-multi-agent-review',
            'error-diagnostics-smart-debug',
            'error-handling-patterns',
            'event-sourcing-architect',
            'event-store-design',
            'executing-plans',
            'fastapi-pro',
            'fastapi-templates',
            'framework-migration-code-migrate',
            'framework-migration-deps-upgrade',
            'framework-migration-legacy-modernize',
            'frontend-design',
            'frontend-dev-guidelines',
            'frontend-developer',
            'frontend-mobile-development-component-scaffold',
            'frontend-mobile-security-xss-scan',
            'frontend-security-coder',
            'frontend-trends-2026',
            'full-stack-scaffold',
            'gcp-cloud-run',
            'git-advanced-workflows',
            'git-collaboration-master',
            'git-pr-workflows-git-workflow',
            'git-pr-workflows-onboard',
            'git-pr-workflows-pr-enhance',
            'git-pushing',
            'github-actions-templates',
            'github-mcp',
            'github-workflow-automation',
            'gitlab-ci-patterns',
            'gitops-workflow',
            'go-concurrency-patterns',
            'godot-gdscript-patterns',
            'helm-chart-scaffolding',
            'html-injection-testing',
            'hybrid-cloud-architect',
            'hybrid-cloud-networking',
            'idor-testing',
            'incident-responder',
            'incident-response-incident-response',
            'incident-response-smart-fix',
            'incident-runbook-templates',
            'javascript-mastery',
            'javascript-pro',
            'javascript-testing-patterns',
            'javascript-typescript-typescript-scaffold',
            'k8s-manifest-generator',
            'k8s-security-policies',
            'kpi-dashboard-design',
            'langchain-architecture',
            'linkerd-patterns',
            'llm-app-patterns',
            'memory-safety-patterns',
            'metasploit-framework',
            'microservices-patterns',
            'ml-pipeline-workflow',
            'mobile-design',
            'mobile-developer',
            'mobile-security-coder',
            'modern-javascript-patterns',
            'modern-web-performance',
            'moodle-external-api-development',
            'multi-cloud-architecture',
            'nextjs-app-router-patterns',
            'nextjs-best-practices',
            'nextjs-react-expert',
            'nextjs-supabase-auth',
            'nodejs-backend-patterns',
            'nodejs-best-practices',
            'notion-template-business',
            'nx-workspace-patterns',
            'observability-monitoring-monitor-setup',
            'observability-monitoring-slo-implement',
            'obsidian-clipper-template-creator',
            'on-call-handoff-patterns',
            'openapi-spec-generation',
            'pci-compliance',
            'pdf-official',
            'penetration-tester-master',
            'pentest-checklist',
            'pentest-commands',
            'performance-engineer',
            'performance-profiling',
            'performance-testing-review-ai-review',
            'performance-testing-review-multi-agent-review',
            'plan-writing',
            'planning-with-files',
            'powershell-windows',
            'pptx-official',
            'pricing-strategy',
            'production-code-audit',
            'projection-patterns',
            'prompt-engineer',
            'prompt-engineering-patterns',
            'python-development-python-scaffold',
            'python-packaging',
            'python-patterns',
            'python-performance-optimization',
            'python-pro',
            'python-testing-patterns',
            'rag-engineer',
            'react-best-practices',
            'react-modernization',
            'react-native-architecture',
            'react-native-best-practices',
            'react-patterns',
            'react-state-management',
            'react-ui-patterns',
            'receiving-code-review',
            'requesting-code-review',
            'rust-async-patterns',
            'screen-reader-testing',
            'search-specialist',
            'security-auditor',
            'security-compliance-compliance-check',
            'security-requirement-extraction',
            'security-scanning-security-dependencies',
            'security-scanning-security-hardening',
            'security-scanning-security-sast',
            'segment-cdp',
            'seo-audit',
            'seo-authority-builder',
            'seo-content-auditor',
            'seo-content-planner',
            'seo-structure-architect',
            'similarity-search-patterns',
            'smtp-penetration-testing',
            'social-content',
            'software-architecture',
            'solidity-security',
            'spark-optimization',
            'sql-injection-testing',
            'sql-optimization-patterns',
            'sqlmap-database-pentesting',
            'ssh-penetration-testing',
            'startup-analyst',
            'startup-business-analyst-business-case',
            'startup-business-analyst-financial-projections',
            'startup-business-analyst-market-opportunity',
            'startup-financial-modeling',
            'startup-metrics-framework',
            'stride-analysis-patterns',
            'systematic-debugging',
            'tailwind-design-system',
            'tailwind-patterns',
            'tdd-master-workflow',
            'tdd-workflow',
            'tdd-workflows-tdd-cycle',
            'tdd-workflows-tdd-green',
            'tdd-workflows-tdd-red',
            'tdd-workflows-tdd-refactor',
            'temporal-python-pro',
            'temporal-python-testing',
            'terraform-specialist',
            'test-automator',
            'test-driven-development',
            'test-fixing',
            'testing-automation-mcp',
            'testing-patterns',
            'typescript-advanced-types',
            'typescript-expert',
            'typescript-pro',
            'ui-ux-designer',
            'unit-testing-test-generate',
            'unity-ecs-patterns',
            'using-git-worktrees',
            'vector-database-engineer',
            'vercel-deploy',
            'vercel-deployment',
            'wcag-audit-patterns',
            'web-design-guidelines',
            'web-performance-optimization',
            'web3-testing',
            'webapp-testing',
            'wordpress-penetration-testing',
            'workflow-automation',
            'workflow-orchestration-patterns',
            'workflow-patterns',
            'writing-plans',
            'xlsx-official',
            'zapier-make-patterns'
        ],
        shared: [
            'i18n-master',
            'design-philosophy',
            'ai-master',
            'api-standards',
            'database-master',
            'design-system',
            'dx-toolkit',
            'metrics',
            'security-armor',
            'testing-master'
        ],
        core: '*',
        scripts: [
            'checklist.py',
            'verify_all.py',
            'translate_workflows.py',
            'verify_shared_modules.js'
        ]
    },
    ultra: {
        rules: '*', // Full Suite
        agents: '*', // Full Suite
        workflows: '*', // Full Suite
        skills: '*', // Full Suite
        shared: '*', // Full Suite
        core: '*',
        scripts: '*'
    }
};

module.exports = MANIFEST;
